### PR TITLE
Add rules: `no-dupe-keys` and `no-reserved-keys`

### PR DESCRIPTION
**Pull request by armano2**
This PR implements rules proposed in #86

**DONE:**
- [x] Create documtation
- [x] Add tests
- [x] Implement rule

### CODE REVIEW

**Comment by armano2** [comment]
@michalsnik its ready to review

**Comment by michalsnik** [comment]
Let's do it one after another. It's really good to see you work so fast and being so committed! 💪 , however we didn't even approved the proposition yet. I hope you don't mind if we'll wait a while until merging all of your PRs? We'd like to focus and gracefully add more and more rules while constantly keeping everything well organized.

**Comment by armano2** [comment]
@michalsnik we are just need them in out project to improve quality of code.

pull requests for js validation should be merged/reviewed after your #69 is going to be merged and conflicts are going to be resolved :dancer:

**Comment by michalsnik** [comment]
I added suggestion in #86

**Comment by armano2** [comment]
@michalsnik i found out that we are not supporting
```
['type']: ....
```
in rules 🗡

**Comment by armano2** [comment]
@michalsnik rule splited to:
- `no-dupe-keys`
- `no-reservered-keys`

File: lib/utils/index.js [hunk 1]
```diff
@@ -384,5 +426,48 @@ module.exports = {
         cb(node.arguments[0])
       }
     }
+  },
+
+  iterateProperties (node, scopeNames, cb) {
```
Comment (michalsnik) [inline lib/utils/index.js]: Please add comments that will describe those new functions

File: lib/rules/no-reservered-keys.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+const RESERVER_NAMES = new Set(require('../utils/vue-reserved.json'))
+const SCOPE_NAMES = new Set(['props', 'computed', 'data', 'methods'])
+
+function create (context) {
+  const reservedNames = RESERVER_NAMES
+  const scopeNames = SCOPE_NAMES
+
+  const options = context.options[0] || {}
+  if (options.reserved) {
```
Comment (michalsnik) [inline lib/rules/no-reservered-keys.js]: You can actually simplify config with:
```
const options = context.options[0] || {}
const reservedKeys = new Set([...RESERVED_KEYS, ...options.reserved || []])
const scopes = new Set([...SCOPE_NAMES, ...options.scopes || []])
```
Plus you can then set `RESERVED_KEYS` and `SCOPE_NAMES` as arrays above.
I changed names of variables so keep it in mind :)

File: lib/rules/no-reservered-keys.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ */
+'use strict'
+
+const utils = require('../utils')
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+const RESERVER_NAMES = new Set(require('../utils/vue-reserved.json'))
```
Comment (michalsnik) [inline lib/rules/no-reservered-keys.js]: `RESERVED`, not `RESERVER`

File: lib/rules/no-reservered-keys.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    })
+  }
+
+  // ----------------------------------------------------------------------
+  // Public
+  // ----------------------------------------------------------------------
+
+  return utils.executeOnVueComponent(context, (obj) => {
+    utils.iterateProperties(obj, scopeNames, (name, node, groupName) => {
+      if (groupName === 'data' && name.substr(0, 1) === '_') {
```
Comment (michalsnik) [inline lib/rules/no-reservered-keys.js]: `name[0]` would also work

File: lib/rules/no-reservered-keys.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        context.report({
+          node: node,
+          message: "Field '{{name}}' which start with _ in data is reserved.",
```
Comment (michalsnik) [inline lib/rules/no-reservered-keys.js]: What do you think about this message: `Keys starting with with "_" are reserved in "data" group.` ?

File: lib/rules/no-reservered-keys.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+          data: {
+            name
+          }
+        })
+      } else if (reservedNames.has(name)) {
+        context.report({
+          node: node,
+          message: "Reserved key '{{name}}'.",
```
Comment (michalsnik) [inline lib/rules/no-reservered-keys.js]: `Key '{{name}}' is reserved`

File: lib/rules/no-reservered-keys.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        })
+      }
+    })
+  })
+}
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent overwrite reserved keys',
```
Comment (michalsnik) [inline lib/rules/no-reservered-keys.js]: I added suggestion for description above

File: lib/utils/index.js (hunk 1 above)
Comment (michalsnik) [inline lib/utils/index.js]: Also, what do you think about calling `scopeNames` -> `groups` instead? Scope is actually the same so being semantically correct we'd call it rather a group. At least I think so

File: lib/utils/index.js [hunk 8]
```diff
@@ -384,5 +426,48 @@ module.exports = {
...
+        } else if (node.value.type === 'ObjectExpression') {
+          this.iterateObjectExpression(node.value, name, cb)
+        } else if (node.value.type === 'FunctionExpression') {
+          this.iterateFunctionExpression(node.value, name, cb)
+        }
+      })
+  },
+
+  iterateArrayExpression (node, groupName, cb) {
+    node.elements.forEach(item => {
```
Comment (michalsnik) [inline lib/utils/index.js]: I'd check type of the passed node in each of those functions first.

File: tests/lib/rules/no-dupe-keys.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      filename: 'test.vue',
+      code: `
+        export default {
+          ...foo(),
+          props: {
+            ...foo(),
+            foo: String
+          },
+          computed: {
+            ...mapGetters({
```
Comment (michalsnik) [inline tests/lib/rules/no-dupe-keys.js]: We don't support checking `mapGetters` and `mapState` case yet right? It's probably something worth considering but only once we have `vuex` environment available in this plugin.

Comment (armano2) [inline tests/lib/rules/no-dupe-keys.js]: @michalsnik we are not supporting it yet, and i'm waiting with it for respond about environments 💃 this is just example with spread operator.

File: lib/rules/no-reservered-keys.js (hunk 2 above)
Comment (armano2) [inline lib/rules/no-reservered-keys.js]: spread operator is not working with node 4

**Comment by armano2** [comment]
@michalsnik suggestions applied

File: lib/utils/index.js [hunk 10]
```diff
@@ -398,5 +440,69 @@ module.exports = {
         cb(node.arguments.slice(-1)[0])
       }
     }
+  },
+
+  iterateProperties (node, groups, cb) {
```
Comment (mysticatea) [inline lib/utils/index.js]: There is the reason you avoid generator functions by callback style?
I like `for-of`.

**Comment by armano2** [comment]
Node 4 :D

File: lib/utils/index.js (hunk 10 above)
Comment (armano2) [inline lib/utils/index.js]: Node 4

**Comment by mysticatea** [comment]
Node 4 supports generators and for-of :smile:

**Comment by armano2** [comment]
Aaaa I was almost 100% sure than it's not supported....

Than I can change it after launch

**Comment by armano2** [comment]
@mysticatea updated

**Comment by mysticatea** [comment]
@armano2 Thank you for prompt update! I'm sorry that I give up today... (AM 1:30 of Monday in my timezone).

File: lib/utils/index.js [hunk 11]
```diff
@@ -273,6 +273,48 @@ module.exports = {
   },
 
   /**
+   * Gets the property name of a given node.
+   * @param {ASTNode} node - The node to get.
+   * @return {string|null} The property name if static. Otherwise, null.
+   */
+  getStaticPropertyName (node) {
```
Comment (michalsnik) [inline lib/utils/index.js]: Can you please add test for this method?

Comment (armano2) [inline lib/utils/index.js]: @michalsnik added

**Pull request merged**